File: MPI/C/sequentialSearchMPI.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <mpi.h>

#define MASTER 0 // Master process
#define END_OF_PROCESSING 0 // End of processing

// COMPILAR
// mpicc sequentialSearchMPI.c -o sequentialSearchMPIc -lm

int* leeArchivo(int* n);
void printArray(int* a, int n);

// PROBAR, AL ENVIAR DATOS, 
// - ES MAS EFICIENTE ENVIAR UNA PARTE DEL ARRAY
// - PONERLO COMO VARIABLE GLOBAL
// - ENVIAR TODO EL ARRAY CON MPI_Bcast()


int main(int argc, char *argv[]){

  	int myrank, numProc;		// RANK Y NUMERO DE PROCESOS
	int tag;					// TAG 
	MPI_Status status;			// MPI STATUS PARA RECIBIR MENSAJES 

    double timeStart, timeEnd;	// CALCULAR TIEMPO DE EJECUCION

    int x=2;
    int* a;                     // ARRAY DINAMICO
    int n,izq,der;              // NUMERO DE ELEMENTOS EN EL ARRAY
    int arrayProc, tamProc;
    int encontrado=-1;


	int sentRows;				/** Number of rows sent */
	int currentRow;				/** Current row being processed */
	int processedRows;			/** Number of currently processed rows */
	int i, j, k;				/** Aux variables */	
		 

    // Init
    tag = 1;
    srand(time(NULL));

    // Init MPI
    MPI_Init (&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &myrank);	
    MPI_Comm_size(MPI_COMM_WORLD, &numProc); // DEVUELVE EL NUMERO DE HILOS AL EJECUTAR EL PROGRAMA

    // CONTROL DE ERRORES
    /*if (){
        if (myrank == MASTER) {
            printf("Wrong configuration for NROWS (%d). SIZE (%d) < numWorkers(%d)*NROWS(%d)\n", NROWS, SIZE, numProc-1, NROWS);
            printf ("At least, each worker must receive %d rows to be processed\n", NROWS);
        }
        MPI_Finalize();
        exit(-1);			
    }*/

		
    // PROCESO master LEE EL ARCHIVO
    if (myrank == MASTER) {
        a=leeArchivo(&n); 
        printf("Padre:\n");
        printArray(a, n);
        // MODIFICAR, POTENCIAS DE 2, LOG2 ...
        // ENVIA TAMAÑO DEL ARRAY QUE VAN A PROCESAR 
        MPI_Bcast (&n, 1, MPI_INT, MASTER, MPI_COMM_WORLD);  
        // ENVIA EL ARRAY ENTERO
        MPI_Bcast (a, n, MPI_INT, MASTER, MPI_COMM_WORLD);  
    }
    else {
        // RECIBE TAMAÑO DEL ARRAY QUE VAN A PROCESAR LOS workers
        MPI_Bcast (&n, 1, MPI_INT, MASTER, MPI_COMM_WORLD);  
        a = (int*)malloc (n*sizeof(int)); 
        
        // RECIBE EL ARRAY ENTERO
        MPI_Bcast (a, n, MPI_INT, MASTER, MPI_COMM_WORLD); 
        printf("Hijo:\n");
        printArray(a, n);          
    }   
            
    // DESPUES DE ENVIAR/RECIBIR LOS DATOS COMIENZA EL TIMER
    timeStart = MPI_Wtime();

		// PROCESO master
		if (myrank == MASTER){
			// INIT
			arrayProc=0;
            // CAMBIAR VALORES
            tamProc=log2(n);

			// DISTRIBUIR INICIALMENTE
			for (i=1; i<numProc; i++) {
				// SEND IZQ
                MPI_Send (&arrayProc, 1, MPI_INT, i, tag, MPI_COMM_WORLD);
                // UPDATE
                arrayProc+=tamProc-1;
                // SEND DER
                MPI_Send (&arrayProc, 1, MPI_INT, i, tag, MPI_COMM_WORLD);
                arrayProc++;							
			}

			// MIENTRAS QUE HAYA ELEMENTOS POR PROCESAR
			while (arrayProc < n) {			
				// RECIBE RESULTADOS DE UNA BUSQUEDA
                printf("PADRE. Esperando\n");
                MPI_Recv (&encontrado, 1, MPI_INT, MPI_ANY_SOURCE, tag, MPI_COMM_WORLD, &status);	
                printf("PADRE. Recibe: %d\n", &encontrado);
                // FIN
                if(encontrado!=-1){
                    printf("Se ha encontrado el elemento en la posicion %d, del array\n", encontrado);
                    timeEnd = MPI_Wtime();
                    //MPI_Abort(MPI_COMM_WORLD, 0);
                }			

				// ENVIAR SIGUIENTES DATOS				
                // SEND IZQ                
                MPI_Send (&arrayProc, 1, MPI_INT, status.MPI_SOURCE, tag, MPI_COMM_WORLD);
                // UPDATE
                if(arrayProc+tamProc>=n) arrayProc=n-1;
                else arrayProc+=tamProc-1;                    
                // SEND DER                
                MPI_Send (&arrayProc, 1, MPI_INT, status.MPI_SOURCE, tag, MPI_COMM_WORLD);
                arrayProc++;	
                //printf("PADRE. arrayProc: %d\n", arrayProc);							
			}

            if(encontrado==-1) {
			    timeEnd = MPI_Wtime();				
                printf("No se ha encontrado el elemento buscado\n");
            }		
            printf("Tiempo de ejecucion: %f\n", timeEnd-timeStart);			
		}		
		else { // workers
            while(encontrado==-1){
                printf("HIJO. Elemento a buscar %d\n",x);
                // RECIBE IZQ
                MPI_Recv (&izq, 1, MPI_INT, MASTER, tag, MPI_COMM_WORLD, &status);                
                // RECIBE DER
                MPI_Recv (&der, 1, MPI_INT, MASTER, tag, MPI_COMM_WORLD, &status);
                printf("HIJO, %d. izq: %d, der: %d\n", myrank, izq, der);
                if(izq>=n) break;
                if(der>=n) der=n-1;
                
                for(;izq<=der&&encontrado==-1;izq++){
                    printf("Valor de a[i]: %d\n",a[izq]);
                    if(a[izq]==x) encontrado=izq;
                }
               
                MPI_Send(&encontrado, 1, MPI_INT, MASTER, tag, MPI_COMM_WORLD);
                printf("HIJO, %d. Envia encontrado: %d\n", myrank, encontrado);                
            }
        
		}
		
		// End MPI environment
  		MPI_Finalize();

    return 0;
}



// cambiar el metodo de seleccionar .txt, usando argv[]
int* leeArchivo(int* n){
    int arrayTam=0;
    int capacidad=10;
    int* a=(int*)malloc(capacidad*sizeof(int));
    if (a == NULL) {
        perror("Error al asignar memoria. 1er");
        exit(EXIT_FAILURE);
    }

    FILE *archivo;
    char nombre_archivo[100];
    int archivoTam;      
    printf("Introduce el nombre del archivo: ");
    scanf("%s", nombre_archivo);
    archivoTam = strlen(nombre_archivo);
    strcat(nombre_archivo, ".txt");

    archivo = fopen(nombre_archivo, "r"); // Modo lectura
    if (archivo == NULL) {
        perror("Error al abrir el archivo");
        exit(EXIT_FAILURE);
    }

    int tmp;
    while (fscanf(archivo, "%d", &tmp) == 1) {        
        if (arrayTam == capacidad) {
            capacidad *= 2;
            a = (int*)realloc(a, capacidad * sizeof(int));
            if (a == NULL) {
                perror("Error al reasignar memoria");
                exit(EXIT_FAILURE);
            }
        }

        a[arrayTam++] = tmp;        
    }

    /*for (int i = 0; i < n; i++) {
        fscanf(archivo, "%d", &a[i]);
    }*/  
    fclose(archivo); 

    *n=arrayTam;
    return a;
}

void printArray(int* a, int n){
    printf("Array:\n");
    for (int i = 0; i < n; i++) {
        printf("%d ", a[i]);
    }
    printf("\n");
}

File: MPI/PYTHON/Busquedas/sequentialSearchMPI.py
from mpi4py import MPI
import sys
import os

"""
COSTE O(N^2)

100000 10;70.1 - 30;56.0 - 15;56.3 - 25;
            MPI (20)    MPI(50)     NORMAL
10000       0.51                    4.5
100000      55          55.9        440.4
1000000     

"""





def main():  
    MASTER = 0              # int.
    INF=sys.maxsize         
    END_OF_PROCESSING=-2    

    timeStart=0.0           # double.
    timeEnd=0.0             
    
    
    # Arrays 
    a=[]                    # int[]. Entrada    
    enc=False                              

    # Variables usadas en el proceso MPI        
    pos=-1                  # int. Variables que se envian y reciben entre los procesos
    val=0                   
    

		    

    # Init MPI.  rank y tag de MPI y el numero de procesos creados (el primero es el master)
    tag=0
    comm=MPI.COMM_WORLD    
    status = MPI.Status()
    myrank=comm.Get_rank()
    numProc=comm.Get_size()
    numWorkers=numProc-1
   
    # CONTROL DE ERRORES
    """if (<Error>){
        if (myrank == MASTER) {
            printf("Wrong configuration for NROWS (%d). SIZE (%d) < numWorkers(%d)*NROWS(%d)\n", NROWS, SIZE, numProc-1, NROWS);
            printf ("At least, each worker must receive %d rows to be processed\n", NROWS);
        }
        MPI_Finalize();
        exit(-1);			
    }"""
		
    # master lee el archivo .txt con el array de entrada
    # cuenta el tamaño del array de entrada e inicializa el array de salida
    if myrank==MASTER: 
        a,n=leeArchivo() 
        val = int(input("Introduce valor a buscar: "))
        
        
    val=comm.bcast(val, root=MASTER)    

    # Comienza el timer una vez inicializado todo
    timeStart=MPI.Wtime()
    
    if myrank==MASTER:
        # Init      
        tamEnviar=n//numWorkers
        workesExtra=n%numWorkers
        izq=0
        der=tamEnviar

        if tamEnviar>=1:
            der+=1
            for i in range(1, workesExtra+1):                                  
                comm.send(a[izq:der], dest=i) 
                comm.send(izq,dest=i)
                #print("Master envia a W",i,a[izq:der])  
                izq+=tamEnviar+1
                der+=tamEnviar+1
            der-=1
            for i in range(workesExtra+1, numWorkers+1):  
                comm.send(a[izq:der], dest=i)   
                comm.send(izq,dest=i)
                izq+=tamEnviar
                der+=tamEnviar
        else :
            aux=numWorkers-workesExtra
            for i in range(1, numWorkers-aux+1):                      
                comm.send(a[(i-1):(i)], dest=i)   
                comm.send((i-1),dest=i)							
            for i in range(numWorkers-aux+1, numWorkers+1):
                comm.send(END_OF_PROCESSING, dest=i)                  
            numWorkers-=aux
      
                            
            
                    

        # Procesa los datos
        while numWorkers>0:	
            rec = comm.recv(source=MPI.ANY_SOURCE, tag=tag,status=status)
            status.Get_source()           

            if rec!=-1: enc=True
            

            # Envia el siguiente valor a procesar    
            comm.send(END_OF_PROCESSING, dest=status.source) 
            numWorkers-=1  
            
        timeEnd = MPI.Wtime()
        
        

        print("Tiempo de ejecucion:", timeEnd-timeStart)
        

        if enc==True: print("Valor,",val,"Encontrado")
        else: print("Valor", val, "No esta en el array")
        #print(b)
    		
    else: # workers
        while(True):
            # Recibe el valor del array a procesar   
            a=comm.recv(source=0) 
            if a==-2: break           
            pos=comm.recv(source=0)
            n=len(a)
            i=0
           

            while i<n:
                if a[i]==val: break
                i+=1
                
            if i==n: comm.send(-1, dest=0)
            else: comm.send(pos+i, dest=0)
            
          
    # End MPI environment
    #MPI.Finalize()

    return 0




def leeArchivo():
    """
    return:
    array: int[].   Array con los enteros leidos
    tam: int.       Tamaño del array leido
    """
    
    tfg_directorio=os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(os.getcwd()))))    
    nombre_fichero=input("Introduce un nombre del fichero: ")    
    path=os.path.join(tfg_directorio, ".Otros","ficheros","No_Ordenados", nombre_fichero+".txt")
    
       
    tam=0    
    array = [] 
    try:        
        with open(path, 'r') as archivo: # modo lectura
            for linea in archivo: # Solo hay una linea                
                numeros_en_linea = linea.split() # Divide por espacios                               
                for numero in numeros_en_linea:
                    array.append(int(numero))
                    tam+=1
    
    except FileNotFoundError:
        print("El archivo '{}' no existe.".format(nombre_fichero+".txt"))
    
    return array, tam


    

def arrayOrdenado(a, n):
    """
    a: int[]    El array a comprobar
    n: int      Tamaño del array
    return.     True or False
    """
    for i in range(1,n):    
        if (a[i]!=a[i-1]+1):return False
    
    return True


main()
